grammar.js

```diff
@@ -1,169 +1,295 @@
+        ['||', PREC.or],
+        ['&&', PREC.and],
+        ['|', PREC.bitwise_or],
+        ['^', PREC.bitwise_xor],
+        ['&', PREC.bitwise_and],
+        ['==', PREC.equality],
+        ['!=', PREC.equality],
+        ['=~', PREC.equality],
+        ['>', PREC.relation],
+        ['<', PREC.relation],
+        ['>=', PREC.relation],
+        ['<=', PREC.relation],
+      ];
+
+      return choice(...table.map(([operator, precedence]) => {
+        return prec.left(precedence, seq(
+          field('left', $.primary_expression),
+          // @ts-ignore
+          field('operator', operator),
+          field('right', $.primary_expression),
+        ));
+      }));
+    },
+
+    unary_expression: $ => prec(PREC.unary, seq(
+      field('operator', choice('+', '-', '!')),
+      field('argument', $.primary_expression),
+    )),
+
+    ternary_expression: $ => prec.left(PREC.ternary, seq(
+      field('condition', $.primary_expression),
+      '?',
+      field('consequence', $.primary_expression),
+      ':',
+      field('alternative', $.primary_expression),
+    )),
+
+    parenthesized_expression: $ => prec(PREC.parenthesized_expression, seq(
+      '(',
+      $.primary_expression,
+      ')',
+    )),
+
+    _literal: $ => choice($.number, $.float, $.boolean, $.string),
+
+    // NOTE: why an optional preceding : before my change?
+    // number: _ => /(?:[0-9]+[.])?[0-9]+/,
+
+    number: _ => /\d+/,
+
+    float: _ => /(\d+(\.\d+)?|\.\d+)/,
+
+    boolean: _ => choice('true', 'false'),
+
+    // Here we tolerate unescaped newlines in double-quoted and
+    // single-quoted string literals.
+    //
+    string: $ => choice(
+      seq(
+        '"',
+        repeat(choice(
+          $.interpolation,
+          $.escape_sequence,
+          $._unescaped_double_string_fragment,
+        )),
+
+        '"',
       ),
-    json_dot_access: $ =>
-      prec.right(
-        9,
-        seq($.simplexpr, field("operator", "."), alias($.ident, $.index))
+      seq(
+        '\'',
+        repeat(choice(
+          $.interpolation,
+          $.escape_sequence,
+          $._unescaped_single_string_fragment,
+        )),
+        '\'',
       ),
-    json_safe_dot_access: $ =>
-      prec.right(
-        9,
-        seq($.simplexpr, field("operator", "?."), alias($.ident, $.index))
+      seq(
+        '`',
+        repeat(choice(
+          $.interpolation,
+          $.escape_sequence,
+          $._unescaped_backtick_fragment,
+        )),
+        '`',
       ),
-    unary_expression: $ =>
+    ),
+
+    interpolation: $ => seq(
+      '${',
+      alias($.primary_expression, $.interpolation_expression),
+      '}',
+    ),
+
+    _escape_sequence: $ =>
       choice(
-        prec.right(8, seq(field("operator", "!"), $.simplexpr)),
-        prec.right(8, seq(field("operator", "-"), $.simplexpr)),
-        prec.right(8, seq(field("operator", "+"), $.simplexpr))
+        prec(2, token.immediate(seq('\\', /[^abfnrtvxu'\"\\\?]/))),
```

Comment:
Why change this to something different than `/./`? If looking at the original parser here: https://github.com/elkowar/eww/blob/6a64a286291456984011c4fb160def106ee55dff/crates/simplexpr/src/parser/lexer.rs#L211

The original parser ignores any char, so this is actually not correct. In fact it just skips any escaped chars at all (doesn't introduce anything special in the AST), but to be able to highlight this, I've introduced a new node `escape_sequence` to be able to specify highlighting for escape sequences. If any of the specified characters should be highlighted differently, this should likely be done with query highlights.

I have also originally thought about copying the escape_sequence from something like javascript, but the parser of `simplexpr` doesn't handle textual encoded unicode characters such as `/u[0-9a-fA-F]{4}/`, it just skips one character (the next after the escaping)

Reply:
I see, I had never taken a look at the original parser, sorry about that!! 

So, if I'm understanding you right, only one-char escape sequences should be highlighted? Like \n, \b, \t, etc..